### DISCUSSION TITLE
Tutorial Error - The toggle switch doesn't change background color

### DISCUSSION THREAD

**Discussion by YaanStart** [iTwin/itwinjs-core#3604] [Q&A]
hi
I am a complete newbie to iTwin.
I'm looking at the second example of the iTwin platform tutorial.

https://developer.bentley.com/tutorials/itwin-viewer-hello-world/

The example was written in iTwin v2, I knew I had to change the package name.
If I change the package name in MyMyFirstUiProvider.tsx and not change App.tsx, it compiles without any problem.

MyFirstUiProvider.tsx
```
import { ColorDef } from '@itwin/core-common';
import { IModelApp } from '@itwin/core-frontend';
import {
     AbstractWidgetProps,
     StagePanelLocation,
     StagePanelSection,
     UiItemsProvider,
} from '@itwin/appui-abstract';
import { ToggleSwitch } from '@itwin/itwinui-react';

import * as React from 'react';

export class MyFirstUiProvider implements UiItemsProvider {
  public readonly id = 'HelloWorldProvider';
  public static toggledOnce: boolean = false;
  public static originalColor: number;

  public provideWidgets(
    stageId: string,
    stageUsage: string,
    location: StagePanelLocation,
    section?: StagePanelSection
  ): ReadonlyArray<AbstractWidgetProps> {
    const widgets: AbstractWidgetProps[] = [];    
    if (
      location === StagePanelLocation.Right &&
      section === StagePanelSection.Start
    )
    {        
      const backgroundColorWidget: AbstractWidgetProps = {        
        id: 'BackgroundColorWidget',
        label: 'Background Color Toggle',
        getWidgetContent() {
          return (
            <ToggleSwitch
              onChange={(toggle) => {
                if (MyFirstUiProvider.toggledOnce === false) {
                  MyFirstUiProvider.originalColor =
                    IModelApp.viewManager.selectedView!.displayStyle.backgroundColor.tbgr;
                  MyFirstUiProvider.toggledOnce = true;
                }

                const color = toggle
                  ? ColorDef.computeTbgrFromString('skyblue')
                  : MyFirstUiProvider.originalColor;
                IModelApp.viewManager.selectedView!.overrideDisplayStyle({
                  backgroundColor: color,
                });
              }}
            />
          );
        },
      };

      widgets.push(backgroundColorWidget);
    }

    return widgets;
  }
}
```

However, when the toggle switch is turned on/off, the background color is changed only once.
Once the background color has been changed, it does not change even if the toggle switch is turned on/off.

![image](https://user-images.githubusercontent.com/105346493/167797303-3396eaac-f6f1-4ea6-9225-2da257027317.png)

![image](https://user-images.githubusercontent.com/105346493/167797522-22bcf2b0-fafb-4b9b-8703-e8d9e158b641.png)

Except for changing the package name, I used the example source as it is.
I wonder why the background color doesn't change according to the toggle switch.

**Comment by ayushibasaiwal** [comment]
Hi @YaanStart , In this tutorial (https://developer.bentley.com/tutorials/itwin-viewer-hello-world)  Component Toggle is used instead of ToggleSwitch which is working perfectly. 
)
The reason behind this:   Considering Toggle, it has an onChange function where it takes a parameter (toggle is named in the tutorial ) which is true when the toggle button is on and false when it is off. But in the case of ToggleSwitch, the parameter which onChange takes is not a boolean but an object. So, the if/else part is not working as per requirement hence switch is not working as in the ToggleSwitch part it is not coming to the else part any time and thus background color is not reverting back to its original color.

Another Workaround:
If we want ToggleSwitch only instead of Toggle then instead of writing if/else using that toggle parameter we can define a separate variable(isToggle in this case) and we can initialize it by true and make it on/off when the toggle is turned on/off. 

Attached code:

import {
  AbstractWidgetProps,
  StagePanelLocation,
  StagePanelSection,
  UiItemsProvider,
  } from "@itwin/appui-abstract";
  import { ColorDef } from "@itwin/core-common";
  import { IModelApp } from "@itwin/core-frontend";
  import { ToggleSwitch } from "@itwin/itwinui-react";
  
  export class SmartDeviceProvider implements UiItemsProvider {
  public readonly id = 'HelloWorldProvider';
  public static toggledOnce: boolean = false;
  public static originalColor: number;
  public static isToggle: boolean= true;
  
  public provideWidgets(
  stageId: string,
  stageUsage: string,
  location: StagePanelLocation,
  section?: StagePanelSection
  ): ReadonlyArray < AbstractWidgetProps> {
  const widgets: AbstractWidgetProps[] = [];
  if (
  location === StagePanelLocation.Right &&
  section === StagePanelSection.Start
  ) {
  const backgroundColorWidget: AbstractWidgetProps = {
  id: 'BackgroundColorWidget',
  label: 'Background Color Toggle',
  getWidgetContent() {
  return (
  <ToggleSwitch
  onChange={(toggle) => {
  if (SmartDeviceProvider.toggledOnce === false) {
  SmartDeviceProvider.originalColor =
  IModelApp.viewManager.selectedView!.displayStyle.backgroundColor.tbgr;
  SmartDeviceProvider.toggledOnce = true;
  }
  let color;
  if(SmartDeviceProvider.isToggle){
  color = ColorDef.computeTbgrFromString('skyblue');
  SmartDeviceProvider.isToggle= false;
  }
  else {
  color = SmartDeviceProvider.originalColor;
  SmartDeviceProvider.isToggle= true;
  }
  IModelApp.viewManager.selectedView!.overrideDisplayStyle({
  backgroundColor: color,
  });
  }}
  />
  );
  },
  };
  widgets.push(backgroundColorWidget);
  }
  return widgets;
  }}


Thanks

**Comment by williamkbentley** [comment]
The tutorial has been updated to 3.x and should work correctly now.
